### ISSUE TITLE
Question: TangentSpace.fit_transform() not adhering sklearn's independence principle?

### ISSUE THREAD

**Issue by kjersbry** [pyRiemann/pyRiemann#114]
Hi!
According to scikit-learn [https://scikit-learn.org/stable/developers/develop.html](url): 

> When fit is called, any previous call to fit should be ignored. In general, calling estimator.fit(X1) and then estimator.fit(X2) should be the same as only calling estimator.fit(X2)

Which means this should also hold for fit_transform(). However, I don't think this is the case with the fit_transform method of TangentSpace? Looking at the implementation of fit():

`def fit(self, X, y=None, sample_weight=None):`
`    '''[...]'''`
`    # compute mean covariance`
`    self.reference_ = mean_covariance(X, metric=self.metric, sample_weight=sample_weight)`
`    # ...`

and then fit_transform():

`def fit_transform(self, X, y=None, sample_weight=None):`
`    '''[...]'''`
`    # compute mean covariance`
`    self._check_reference_points(X)`
`    self.reference_ = mean_covariance(X, metric=self.metric, sample_weight=sample_weight)`
`    # ...`

If fit() has been called with data with one certain format (e.g. 7 channels), and fit_transform is later called with another format (e.g. 5 channels), it crashes with the error  ValueError('Data must be same size of reference point.') from `self._check_reference_points(X)`, because fit_transform first checks self.reference_. But self.reference_ was set while fit() was called with the first dataset. I believe running `  self._check_reference_points(X)` first thing in fit_transform(), makes fit_transform() uncompatible with the sklearn principle mentioned above that all calls to fit() should be independent? Maybe there is some other reason for this, to make it compatible with something else?

That being said, I got this error in what I assume is an edge case (and was able to resolve it), and I guess this error only appears in edge cases since you seldom will use a single TS object on different types of data. So I'm mainly just pointing out that TS.fit_transform() possibly doesn't adhere with the sklearn fit() principle, and I'm curious if there's a reason for this or if it should be corrected. :)

**Comment by alexandrebarachant** [comment]
Thanks for raising the issue. 

To be perfectly honest, it can't recall exactly why this is done this way. 

The tangent space `transform` is already breaking aways from sklearn patterns in case of `tsupdate=True` (this will effectively fit again the reference point upon calling transform). 

But i think you are right, for the fit_transform purpose, i have hard time to see a usecase where you would need to reuse past tangent point, and its overridden anyway on the next line. We can definitely remove the check from the fit_transform.

**Comment by agramfort** [comment]
@kjersbry can you send a PR to fix the fit behavior that you could fit twice without memory of the first fit?

**Comment by kjersbry** [comment]
I just quickfixed it by making a new pipeline object every time, so I'm afraid I don't have a fix for fitting twice, sorry!

**Comment by agramfort** [comment]
we can always hack our way forward. The question is how to make sure nobody
hits this bug again.
If we all think like this we are collectly slower...

>

**Comment by kjersbry** [comment]
Ok, I'll send a PR where the check is removed in fit_transform. It'll be my first pull request, so I'll be happy to hear if I should have done something differently :)

**Comment by sylvchev** [comment]
Thanks! Do not hesitate to ask, if you are stuck or if you need help :)
